hcl/hcl.go

```diff
@@ -873,16 +879,135 @@ func CopyBody(target *hclwrite.Body, src *hclsyntax.Body, eval PartialEvaluator)
-		if block.Body == nil {
+		err := appendBlock(target, block, eval)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func appendBlock(target *hclwrite.Body, block *hclsyntax.Block, eval Evaluator) error {
+	if block.Type == "tm_dynamic" {
+		return appendDynamicBlock(target, block, eval)
+	}
+
+	targetBlock := target.AppendNewBlock(block.Type, block.Labels)
+	if block.Body != nil {
+		err := CopyBody(targetBlock.Body(), block.Body, eval)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func appendDynamicBlock(target *hclwrite.Body, block *hclsyntax.Block, eval Evaluator) error {
+	attrs := block.Body.Attributes
+
+	errs := errors.L()
+	if len(block.Labels) != 1 {
+		errs.Append(errors.E(ErrTerramateSchema,
+			block.LabelRanges, "tm_dynamic requires a label"))
+	}
+	genBlockType := block.Labels[0]
+	var genContentBlock *hclsyntax.Block
+
+	for _, b := range block.Body.Blocks {
+		if b.Type != "content" {
+			errs.Append(errors.E(ErrTerramateSchema,
+				b.TypeRange, "unrecognized block %s", b.Type))
+
 			continue
 		}
-		if err := CopyBody(targetBlock.Body(), block.Body, eval); err != nil {
+
+		if genContentBlock != nil {
+			errs.Append(errors.E(ErrTerramateSchema, b.TypeRange,
+				"multiple definitions of the `content` block"))
+
+			continue
+		}
+
+		genContentBlock = b
+	}
+
+	if genContentBlock == nil {
+		errs.Append(errors.E(ErrTerramateSchema, block.Body.Range(),
+			"`content` block not defined"))
+	}
+
+	if err := errs.AsError(); err != nil {
+		return err
+	}
+
+	iterator := genBlockType
+	iteratorAttr, ok := attrs["iterator"]
+	if ok {
+		iteratorTraversal, diags := hcl.AbsTraversalForExpr(iteratorAttr.Expr)
+		if diags.HasErrors() {
+			return errors.E(diags, ErrInvalidDynamicIterator,
+				"failed to traverse expression")
+		}
+		if len(iteratorTraversal) != 1 {
+			return errors.E(iteratorAttr.Expr.Range(),
+				ErrInvalidDynamicIterator,
+				"dynamic iterator must be a single variable name.")
+		}
+		iterator = iteratorTraversal.RootName()
+	}
+
+	var labels []string
+	labelsAttr, ok := attrs["labels"]
+	if ok {
+		labelsVal, err := eval.Eval(labelsAttr.Expr)
+		if err != nil {
+			return errors.E(err, "failed to evaluate the `labels` attribute")
+		}
+
+		err = assignSet("labels", &labels, labelsVal)
+		if err != nil {
 			return err
 		}
 	}
 
-	return nil
+	forEachAttr, ok := attrs["for_each"]
+	if !ok {
+		return errors.E(block.Body.Range(),
+			ErrTerramateSchema,
+			"tm_dynamic requires a `for_each` attribute")
+	}
+
+	forEachVal, err := eval.Eval(forEachAttr.Expr)
+	if err != nil {
+		return errors.E(forEachAttr.Expr.Range(),
+			"evaluting `for_each` expression")
+	}
+
+	if !forEachVal.CanIterateElements() {
+		return errors.E(forEachAttr.Expr.Range(),
+			"expression value of type %s cannot be iterated",
+			forEachVal.Type().FriendlyName())
+	}
+
+	var tmDynamicErr error
+	forEachVal.ForEachElement(func(key, value cty.Value) (stop bool) {
```

Comment:
```suggestion
	forEachVal.ForEachElement(func(key, value cty.Value) (bool) {
```

The named return is not used, so it seemed a little strange, but if it is about documenting the return value all good.

Reply:
yes it was just to document it because my intuition expected the opposite, like, returning true would be "continue iterating" instead of "stop iterating"...